extensions/wikia/CategoryPage3/CategoryPage3.class.php

```
@@ -0,0 +1,194 @@
+		$html .= $this->getHTMLForMembersHeader( $engine );
+		$html .= $this->getHTMLForMembers( $engine );
+		$html .= $this->getHTMLForPagination( $engine );
+
+		return $html;
+	}
+
+	private function getHTMLForTrendingPages( Engine $engine ): string {
+		if ( empty( $this->trendingPages ) ) {
+			return '';
+		}
+
+		return $engine->clearData()
+			->setData( [
+				'trendingPages' => $this->trendingPages
+			] )
+			->render( 'trending-pages.php' );
+	}
+
+	private function getHTMLForMembersHeader( Engine $engine ): string {
+		return $engine->clearData()
+			->setData( [
+				'alphabetShortcuts' => $this->getAlphabetShortcuts(),
+				'title' => $this->getTitle(),
+				'totalNumberOfMembers' => $this->model->getTotalNumberOfMembers()
+			] )
+			->render( 'members-header.php' );
+	}
+
+	private function getHTMLForMembers( Engine $engine ): string {
+		return $engine->clearData()
+			->setData( [
+				'hasSingleMember' => count( $this->model->getMembers() ) === 1,
+				'membersGroupedByChar' => $this->model->getMembersGroupedByChar()
+			] )
+			->render( 'members.php' );
+	}
+
+	private function getHTMLForPagination( Engine $engine ): string {
+		return $engine->clearData()
+			->setData( [
+				'pagination' => $this->model->getPagination()
+			] )
+			->render( 'pagination.php' );
+	}
+
+	public static function getSrcset( string $url ): string {
+		// Based on how much space there is on different breakpoints
+		$widths = [ 160, 200 ];
+		$srcSetItems = [];
+
+		foreach ( $widths as $width ) {
+			$thumb = self::getThumbnailUrlForWidth( $url, $width );
+			$srcSetItems[] = "${thumb} ${width}w";
+		}
+
+		return implode( ',', $srcSetItems );
+	}
+
+	private static function getThumbnailUrlForWidth( string $url, int $width ): string {
+		// Keep 4:3 ratio
+		$height = floor( $width / 4 * 3 );
+
+		try {
+			$url = VignetteRequest::fromUrl( $url )
+				->topCrop()
+				->width( $width )
+				->height( $height )
+				->url();
+		} catch ( Exception $e ) {
```

It looks like it only explicitly throws `InvalidArgumentException`, so being more specific here might be good to avoid potentially hiding other exceptions that we would want to know happened.

---

`->url()` throws `\Exception`:
https://github.com/Wikia/app/blob/6c60bc14c47ed41d2d72683c137e3c1f9acbaf27/lib/Wikia/src/Vignette/UrlGenerator.php#L283-L285